CI reorg - run multiple builds in same checkout

evelikov opened the pull request:
A quick and dirty attempt of my ideas at https://github.com/kmod-project/kmod/issues/320#issuecomment-2919024212

Note, this is _far_ from complete, but I'm posting for early feedback. Also I'm not a huge fan of the duplicated boilerplate in each step - a few possible solutions:
 - create separate jobs (as opposed to one) annotate dependencies - 32bit gcc test, needs 32bit gcc build, etc
 - make the boilerplate a function, store that in GITHUB_ENV and reuse

Any ideas and feedback are appreciated o/

lucasdemarchi commented:
I think one possible thing to do is to split the container setup from the test. Defining one job to update the container and another to build/test the code. Updating the containers could be done ~ weekly and on PRs they are only updated if they are touched.

lucasdemarchi commented:
splitting the container from the actual test was what I did for git-pile:
https://github.com/git-pile/git-pile/blob/master/.github/workflows/build.yml#L72
and https://github.com/git-pile/git-pile/blob/master/.github/workflows/container.yml although I didn't update the container regularly.

lucasdemarchi reviewed (commented):
I'm ok going this route. Another idea to reduce the duplication:

Add a step that simply creates a function `for_each_config` in a .common-functions.sh file that you then source on each step. Each step now can do something like this:

```
function setup() {
    if [[ "$matrix[bits]" == "32" ]]; then
        setup_options=...
    else
        setup_options=""
    fi
    meson setup ...
}

. .common-functions.sh
for_each_config "<matrix options>" setup
```

lucasdemarchi commented on .github/workflows/main.yml:
```diff
@@ -18,99 +18,47 @@ permissions:
-            container: 'debian:bullseye-slim'
+          - container: 'archlinux:multilib-devel'
+            # fails on LD_PRELOAD
+            skip_test_32: 'true'
```
I mean... you are setting skip_test_32 to true, but later you are only checking "skip", no "skip_test_32"

lucasdemarchi replied:
ah... never mind.... skip=32 is to skip **building** on 32 bits.

evelikov committed:
ci: use meson bla -C builddir/

Use builddir/ as indicated in our README, which makes it obvious that
what the argument is. While there, remove a few instances of mkdir/cd -
meson and ninja can handle it themselves.

Signed-off-by: Emil Velikov <emil.l.velikov@gmail.com>

evelikov committed:
ci: fold 32 and 64 bit jobs together

We currently construct a single container for each distro, which is
capable of running both 32 and 64 bit builds.

In addition, since we're meson only (which allows only out-of-tree
builds) we can do multiple builds for a given checkout.

As result we hammer the docker registry and distribution/package servers
a bit less.

Notes:
 - 32bit distcheck is disabled - not particularly interesting
 - 32bit tests are disabled for everyone - they are consistently failing

Signed-off-by: Emil Velikov <emil.l.velikov@gmail.com>

evelikov committed:
ci: fold gcc/clang combinations together

Instead of listing each combination on per-compiler basis, assume we
want both unless annotated otherwise.

Signed-off-by: Emil Velikov <emil.l.velikov@gmail.com>

evelikov committed:
ci: install and use bash across the board

The upcoming CI rework will hard depend on some bash constructs. While
theoretically possible to make them dash compliant, I'm not sure it's
worth the effort. Just install bash and carry on.

Signed-off-by: Emil Velikov <emil.l.velikov@gmail.com>

lucasdemarchi commented:
Applied, thanks.
